src/core/server/saved_objects/migrations/kibana/kibana_migrator.ts

```diff
@@ -230,6 +234,7 @@ export class KibanaMigrator {
    * @returns `doc` with all registered migrations applied.
    */
   public migrateDocument(doc: SavedObjectUnsanitizedDoc): SavedObjectUnsanitizedDoc {
+    this.documentMigrator.prepareMigrations();
     return this.documentMigrator.migrate(doc);
   }
```

Comment:
`migrateDocument` is called during the SOR's `create`, `bulkCreate` and `incrementCounter` calls.

Ideally, we would avoid calling `documentMigrator.prepareMigrations` everytime we are invoking these APIs.

Theoretically, `migrateDocument` should not be called before the SO migration. Maybe asserting/throwing if `KibanaMigrator.migrateDocument` is called before would be better?

Or maybe `documentMigrator.prepareMigrations` should no-op after the initial call? That may be more error prone though, if for some reason `KibanaMigrator. migrateDocument` is invoked before core's start lifecycle (which should not be the case atm)

wdyt @rudolf?

Reply:
I initially had a guard to only invoke `prepareMigrations` once but I thought this might lead to unexpected edge cases in case the logic changes in the future.

I think throwing an error makes the most sense here since we have a known order of execution of things here.

Reply:
> Maybe asserting/throwing if KibanaMigrator.migrateDocument is called before would be better?

yeah this is on the hot path for performance, so that's a good suggestion